File: ruby/circular-buffer/circular_buffer.rb
class CircularBuffer
  def initialize(size)
    @buffer = Array.new(size)
    @head = 0
    @tail = 0
  end

  def size
    @tail - @head
  end

  def empty?
    size == 0
  end

  def full?
    size == @buffer.size
  end

  def read
    raise BufferEmptyException if empty?

    val = @buffer[@head % @buffer.size]
    @head += 1

    return val
  end

  def write(x)
    raise BufferFullException if full?

    @buffer[@tail % @buffer.size] = x
    @tail += 1
  end

  def write!(x)
    begin
      write(x)
    rescue BufferFullException
      @buffer[@tail % @buffer.size] = x

      @head += 1
      @tail += 1
    end
  end

  def clear
    @head = @tail = 0
  end

  class BufferEmptyException < StandardError; end
  class BufferFullException < StandardError; end
end

File: csharp/circular-buffer/CircularBuffer.cs
using System;

public class CircularBuffer<T>
{
    int Capacity, Head, Tail;
    T[] Buffer;

    public CircularBuffer(int capacity)
    {
        Capacity = capacity;
        Head = Tail = 0;

        Buffer = new T[Capacity];
    }

    public int Size
    {
        get { return Tail - Head; }
    }

    public bool Empty()
    {
        return Size == 0;
    }

    public bool Full()
    {
        return Size == Capacity;
    }

    public T Read()
    {
        if (Empty())
            throw new InvalidOperationException();

        return Buffer[Head++ % Capacity];
    }

    public void Write(T value)
    {
        if (Full())
            throw new InvalidOperationException();

        Buffer[Tail++ % Capacity] = value;
    }

    public void Overwrite(T value)
    {
        try {
            Write(value);
        } catch (InvalidOperationException) {
            Buffer[Tail % Capacity] = value;

            Head += 1;
            Tail += 1;
        }
    }

    public void Clear()
    {
        Head = Tail = 0;
    }
}
